### PR TITLE
Add KMM preflightvalidationocp unit-tests

### PR DESCRIPTION
**Pull request by cdvultur**

### CODE REVIEW

File: pkg/kmm/preflightvalidationocp_test.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		namespace           string
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			name:                "test",
+			namespace:           "testns",
+			expectedError:       false,
+			addToRuntimeObjects: true,
```
Comment (achuzhoy) [inline pkg/kmm/preflightvalidationocp_test.go:161]: expectedErrorText  should be explicitly set as well.

Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:161]: Fixed ! Thanks !

File: pkg/kmm/preflightvalidationocp_test.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+		testBuilder := buildPreflightTestBuilderWithFakeObjects(runtimeObjects)
+
+		builderResult, err := testBuilder.Create()
+		assert.Nil(t, err)
+		assert.NotNil(t, builderResult)
+		assert.Equal(t, defaultPreflightName, builderResult.Definition.Name)
+		assert.Equal(t, defaultPreflightNamespace, builderResult.Definition.Namespace)
+	}
+}
+func TestPreflightUpdate(t *testing.T) {
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:252]: Please refer to:
https://github.com/openshift-kni/eco-goinfra/blob/c2a8833ad5034f4d11c21752928abe93e476350c/pkg/argocd/argocd_test.go#L244

File: pkg/kmm/preflightvalidationocp_test.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+}
+func TestPreflightExists(t *testing.T) {
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:255]: Please refer: https://github.com/openshift-kni/eco-goinfra/blob/c2a8833ad5034f4d11c21752928abe93e476350c/pkg/argocd/argocd_test.go#L165

File: pkg/kmm/preflightvalidationocp_test.go [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+		exists        bool
+		expectedError error
+	}{
+		{
+			exists:        true,
+			expectedError: nil,
+		},
+		{
+			exists:        false,
+			expectedError: errors.New("preflightvalidationocp cannot be deleted because it does not exist"),
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:269]: Please fix Delete function. We don't return error if object doesn't exist.

File: pkg/kmm/preflightvalidationocp_test.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+		_, err := testBuilder.Delete()
+		assert.Equal(t, testCase.expectedError, err)
+
+		if testCase.expectedError == nil {
+			assert.Nil(t, testBuilder.Object)
+		}
+	}
+}
+
+func TestPreflightGet(t *testing.T) {
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:295]: Please refer to: pkg/argocd/argocd_test.go

Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:355]: Updated. thanks

File: pkg/kmm/preflightvalidationocp_test.go (hunk 4 above)
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Done

File: pkg/kmm/preflightvalidationocp_test.go (hunk 3 above)
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:301]: Done, Thanks

File: pkg/kmm/preflightvalidationocp_test.go (hunk 2 above)
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:259]: Done, Thanks

File: pkg/kmm/preflightvalidationocp_test.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+		}
+	}
+}
+
+func TestPreflightCreate(t *testing.T) {
+	testCases := []struct {
+		addToRuntimeObjects bool
+	}{
+		{
+			addToRuntimeObjects: true,
```
[the remark this answers is not in the document]
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Done

File: pkg/kmm/preflightvalidationocp_test.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+			testSettings   *clients.Settings
+		)
+
+		testPreflight := generatePreflight(testCase.name, testCase.namespace)
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testPreflight)
+		}
+
+		testSettings = clients.GetTestClients(clients.TestClientParams{
```
[the remark this answers is not in the document]
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Updated

File: pkg/kmm/preflightvalidationocp_test.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func TestPullPreflightValidationOCP(t *testing.T) {
+	testCases := []struct {
+		name                string
+		namespace           string
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
[the remark this answers is not in the document]
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:154]: Added

File: pkg/kmm/preflightvalidationocp_test.go [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+			assert.NotNil(t, testBuilder)
+			assert.Equal(t, testCase.expectedPreflight.Name, testBuilder.Definition.Name)
+			assert.Equal(t, testCase.expectedPreflight.Namespace, testBuilder.Definition.Namespace)
+		} else {
+			assert.Equal(t, testCase.expectedErr, testBuilder.errorMsg)
+		}
+	}
+}
+
+func TestPreflightWithReleaseImage(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Done

[the remark this answers is not in the document]
Comment (kononovn) [reply pkg/kmm/preflightvalidationocp_test.go]: All function should have prefix TestPreflightValidation*. Please fix

File: pkg/kmm/preflightvalidationocp_test.go [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+		{
+			name:                "test",
+			namespace:           "testns",
+			expectedError:       true,
+			addToRuntimeObjects: false,
+			expectedErrorText:   "preflightvalidation 'apiClient' cannot be empty",
+			client:              false,
+		},
+		{
+			name:                "test",
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:166]: Something is wrong here.
This is exactly the same as last test case but in last test case you have error.

File: pkg/kmm/preflightvalidationocp_test.go [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+		}
+	}
+}
+
+func TestPreflightCreate(t *testing.T) {
+	testCases := []struct {
+		testPreflight *PreflightValidationOCPBuilder
+		expectedError error
+	}{
+		{
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:239]: Please add test case when object already exist on cluster.

Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go:226]: Updated the test and refactored the ResourceVersion part

File: pkg/kmm/preflightvalidationocp_test.go (hunk 10 above)
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Indeed. The `expectedError` and `expectedErrorText` were also misleading. 

Fixed it. Thanks !

File: pkg/kmm/preflightvalidationocp_test.go [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+	assert.Equal(t, "", testBuilder.errorMsg)
+
+	testBuilder.WithOptions(func(builder *PreflightValidationOCPBuilder) (*PreflightValidationOCPBuilder, error) {
+		return builder, fmt.Errorf("error")
+	})
+
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func TestPreflightValidationPullPreflightValidationOCP(t *testing.T) {
```
Comment (kononovn) [inline pkg/kmm/preflightvalidationocp_test.go:148]: Please change to TestPullPreflightValidationOCP

File: pkg/kmm/preflightvalidationocp_test.go [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	defaultPreflightName      = "preflight"
+	defaultPreflightNamespace = "preflightns"
+)
+
+func TestPreflightValidationBuilder(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cdvultur) [reply pkg/kmm/preflightvalidationocp_test.go]: Updated! Thanks!

**Review by kononovn — APPROVED** [review]
/lgtm

**Pull request merged**